### ISSUE TITLE
Read PE via stream/buffer

### ISSUE THREAD

**Issue by Symbai** [Washi1337/AsmResolver#115]
Can we get a way to read a PE File via stream/buffer without reading all bytes at once? Some PE files are ~400mb, its insane to read them all into memory when we're looking into the PE header only.

**Comment by Washi1337** [comment]
Yes this is something I was thinking of before and could be very beneficial for a lot of use-cases of the library. It would translate in a different implementation of the `IBinaryStreamReader` interface. 

One challenge with this is that a lot of the initialization of properties is done lazily. By "forking" the underlying `IBinaryStreamReader` object, and letting these forks point to a different chunk of the input file, AsmResolver only processes the data stored in the file when it is requested. 

With a byte array as underlying data source, this is easily implemented, since byte arrays provide full random access at any point in time, and multiple threads can access different parts of the same array at once. With streams however, it is more complicated, since we cannot just copy the reference to the same input stream to new forks of the reader, and start calling the `Seek` method / set the `Position` property method without running into potential concurrency issues. We would need some kind of locking for this, which could potentially impact performance. 

If you have any suggestions on how to implement such a thing, I'd be more than happy to hear :)

**Comment by Symbai** [comment]
Other libraries are using BinaryReader in combination with File.OpenRead stream. And are indeed seeking through it. But I don't experienced any issue with that. Its very fast and uses a lot less memory. Cannot say something about concurrency issue as they aren't multi threaded.

If thread safety is important, MemoryMappedFile would be another choice. It won't read the whole file from disk and it's thread safety.

**Comment by Washi1337** [comment]
Yea it is not so much the underlying binary reader that has threading issues. It is more about how AsmResolver uses the reader object that might result in threading issues. For example, if two threads attempt to access 2 different properties that are lazily initialized with the same binary reader object, then both threads would attempt to seek through the stream before reading the data, resulting in a data race for the current position of the underlying stream object.

`MemoryMappedFile` sounds like a very good alternative, since it seems to be able to create random-access views on the file using `CreateViewAccessor`. Would have to do some reading on how that works exactly though, and see if it works across different platforms (as AsmResolver targets .NET standard 2.0).

Just thinking out loud to see what kind of changes it would require: Luckily, there is not too much code relying on the actual `Fork` method of the `IBinaryStreamReader` anymore as it used to be in earlier versions of the library. Most of the code just relies on implementations of `IReadableSegment`. Currently `SerializedPEFile` (the default PE file implementation for reading existing PE files) just initializes for each section a simple `DataSegment` (which are essentially just wrappers around `byte[]`). We could perhaps change the implementation there.

**Comment by Symbai** [comment]
MemoryMappedFile is in [.NET Standard 2 and in .NET Core 1.0 -> .NET 5](https://docs.microsoft.com/en-us/dotnet/api/system.io.memorymappedfiles.memorymappedfile?view=net-5.0#applies-to). So it should work across all platforms.

**Comment by Washi1337** [comment]
Implemented in 4.6.0.
